src/clj/puppetlabs/http/client/async.clj

```diff
@@ -117,11 +124,68 @@
       (.setAs (clojure-response-body-type->java opts))
       (.setBody (:body opts))
       (.setDecompressBody (clojure.core/get opts :decompress-body true))
-      (.setHeaders (:headers opts))))
+      (.setHeaders (:headers opts))
+      (.setMetricId (parse-metric-id opts))))
+
+(schema/defn get-mean :- schema/Num
+  [timer :- Timer]
+  (->> timer
+       .getSnapshot
+       .getMean
+       (.toMillis TimeUnit/NANOSECONDS)))
+
+(defn get-metric-data
+  [timer metric-id]
+  (let [count (.getCount timer)
+        mean (get-mean timer)
+        aggregate (* count mean)]
+    {:count count
+     :mean mean
+     :aggregate aggregate
+     :metric-id metric-id}))
+
+(defn get-metrics-data
+  [timers]
+  (reduce (fn [acc [metric-id timer]]
+            (assoc acc metric-id (get-metric-data timer metric-id)))
+          {} timers))
 
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 ;;; Public
 
+(schema/defn ^:always-validate get-client-metrics :- (schema/maybe common/Metrics)
+  "Returns the http client-specific metrics from the metric registry."
+  ([metric-registry :- common/OptionalMetricRegistry]
```

Comment:
Since this is used for the implementation of the protocol function `get-client-metrics`, this could be called on a client that didn't have a metrics registry registered. Maybe this should be moved out of the "public" section? I'm not really clear on what exactly should go where.